### PR TITLE
Directory instance in run

### LINKED ISSUE #81: Directory instance in container run methods
Add directory parameter to run methods (instance of Directory) - it is automatically mounted to container at start

### PR DESCRIPTION
**Pull request by lachmanfrantisek**
Signed-off-by: lachmanfrantisek <flachman@redhat.com>

Resolves #81 

What do you think about this approach?

- [x] add to `run_via_binary_in_foreground()`
- [x] unit tests for `Volume` class
- [x] integration tests
- [x] example and/or documentation (added by Tomas)

### CODE REVIEW

File: conu/backend/docker/image.py [hunk 1]
```diff
@@ -254,6 +256,13 @@ def run_via_binary(self, run_command_instance=None, *args, **kwargs):
         run_command_instance.image_name = self.get_id()
         run_command_instance.options += ["-d"]
 
+        if volume:
+            if not isinstance(volume, list):
+                volume = [volume]
+            volumes = [Volume.create_from_tuple(v) for v in volume]
+            for v in volumes:
+                run_command_instance.options += ["-v", v.cmd_option]
```
Comment (dhodovsk) [inline conu/backend/docker/image.py]: There is no cmd_option in Volume. 
str(v)?
```
>>> import conu
>>> cont = conu.DockerImage('fedora').run_via_binary(volume=(conu.Directory('/usr/bin'), "/mountpoint", "Z"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dhodovsk/repos/github.com/fedora-modularity/conu/conu/backend/docker/image.py", line 264, in run_via_binary
    run_command_instance.options += ["-v", v.cmd_option]
AttributeError: 'Volume' object has no attribute 'cmd_option'
```

File: conu/backend/docker/image.py [hunk 2]
```diff
@@ -238,12 +239,13 @@ def _run_container(self, run_command_instance, callback):
...
         return container_id, response
 
-    def run_via_binary(self, run_command_instance=None, *args, **kwargs):
+    def run_via_binary(self, run_command_instance=None, volume=None, *args, **kwargs):
         """
         create a container using this image and run it in background;
         this method is useful to test real user scenarios when users invoke containers using
         binary
 
+        :param volume: tuple or list of tuples inform: target | source,target | source,target,mode
```
Comment (dhodovsk) [inline conu/backend/docker/image.py]: s/inform/in form/
Also usage example in docstring would be helpful.

File: conu/backend/docker/image.py [hunk 3]
```diff
@@ -238,12 +239,13 @@ def _run_container(self, run_command_instance, callback):
             container_id = fd.read()
         return container_id, response
 
-    def run_via_binary(self, run_command_instance=None, *args, **kwargs):
+    def run_via_binary(self, run_command_instance=None, volume=None, *args, **kwargs):
```
Comment (dhodovsk) [inline conu/backend/docker/image.py]: Please add this also to `run_via_binary_in_foreground()`

File: conu/backend/docker/image.py (hunk 1 above)
Comment (lachmanfrantisek) [reply conu/backend/docker/image.py]: omg, I've forgotten to push this.

File: conu/backend/docker/image.py [hunk 4]
```diff
@@ -238,12 +239,16 @@ def _run_container(self, run_command_instance, callback):
...
```
Comment (TomasTomecek) [inline conu/backend/docker/image.py]: I propose to name it `volumes` so it's clear it's suppose to be a list

File: conu/backend/docker/image.py [hunk 5]
```diff
@@ -238,12 +239,16 @@ def _run_container(self, run_command_instance, callback):
...
         """
         create a container using this image and run it in background;
         this method is useful to test real user scenarios when users invoke containers using
         binary
 
+        :param volume: tuple or list of tuples in the form: target | source,target | source,target,mode
```
Comment (TomasTomecek) [inline conu/backend/docker/image.py]: I vote for more verbose description written as python code, something like:
```
 * `("/path/to/directory", )`
 * `("/host/path", "/container/path")`
```
please also make sure sphinx renders this nicely

File: conu/backend/docker/image.py [hunk 6]
```diff
@@ -254,6 +259,13 @@ def run_via_binary(self, run_command_instance=None, *args, **kwargs):
         run_command_instance.image_name = self.get_id()
         run_command_instance.options += ["-d"]
 
+        if volume:
+            if not isinstance(volume, list):
```
Comment (TomasTomecek) [inline conu/backend/docker/image.py]: so this the reason for the name `volume`

File: conu/utils/filesystem.py [hunk 7]
```diff
@@ -222,3 +222,48 @@ def __unicode__(self):
...
+        self.mode = mode
+
+    def __str__(self):
+        """
+        Cmd option representing the volume
+        :return:
+        """
+        result = self.target
+        if self.source:
+            result = "{}:{}".format(self.source.path, result)
```
Comment (TomasTomecek) [inline conu/utils/filesystem.py]: what if `self.source` is a string?

File: conu/utils/filesystem.py [hunk 8]
```diff
@@ -222,3 +222,48 @@ def __unicode__(self):
...
+            return Volume(target=volume)
+        elif len(volume) == 2:
+            return Volume(source=volume[0],
+                          target=volume[1])
+        elif len(volume) == 3:
+            return Volume(source=volume[0],
+                          target=volume[1],
+                          mode=volume[2])
+        else:
+            raise ConuException("Cannot create volume instance.")
```
Comment (TomasTomecek) [inline conu/utils/filesystem.py]: some debug output would be very helpful saying why it failed

File: conu/utils/filesystem.py (hunk 7 above)
Comment (lachmanfrantisek) [reply conu/utils/filesystem.py]: It is converted to `Directory` in `__init__`:
```
        if source and not isinstance(source, Directory):
            self.source = Directory(path=source)
```

File: conu/backend/docker/image.py (hunk 5 above)
Comment (lachmanfrantisek) [reply conu/backend/docker/image.py]: Done.

File: conu/backend/docker/image.py (hunk 4 above)
Comment (lachmanfrantisek) [reply conu/backend/docker/image.py]: Done.

**Comment by lachmanfrantisek** [comment]
@TomasTomecek 

>  example and/or documentation (added by Tomas)

Is this enough?

```
:param volumes: tuple or list of tuples in the form:

    * `("/path/to/directory", )`
    * `("/host/path", "/container/path")`
    * `("/host/path", "/container/path", "mode")`
    * `(conu.Directory('/host/path'), "/container/path")` (source can be also Directory instance)
```

File: conu/utils/filesystem.py (hunk 8 above)
Comment (lachmanfrantisek) [reply conu/utils/filesystem.py]: Done.

File: conu/backend/docker/image.py (hunk 3 above)
Comment (lachmanfrantisek) [reply conu/backend/docker/image.py]: Done.

File: conu/utils/filesystem.py (hunk 7 above)
Comment (TomasTomecek) [reply conu/utils/filesystem.py]: /facepalm

(I missed that)

**Comment by TomasTomecek** [comment]
Changes LGTM, please add/modify one example so users know how to use this and let's ship!

**Comment by lachmanfrantisek** [comment]
@TomasTomecek  I've changed the webserver example to use this feature.

File: docs/source/examples/webserver.py [hunk 9]
```diff
@@ -44,10 +44,12 @@ def run_container(backend, local_dir):
     # the command to run in a container
     command = ["python3", "-m", "http.server", "--bind", "0.0.0.0", "%d" % port]
     # additional options passed to `run` command
-    additional_opts = ["-v", "%s:/webroot" % local_dir, "-w", "/webroot"]
+    additional_opts = ["-w", "/webroot"]
 
     # let's run the container (in the background)
-    container = image.run_via_binary(command=command, additional_opts=additional_opts)
+    container = image.run_via_binary(command=command,
+                                     volumes=(local_dir, "/webroot"),
```
Comment (TomasTomecek) [inline docs/source/examples/webserver.py]: \o/ much nicer

**Comment by lachmanfrantisek** [comment]
Thank both of you!

**Pull request merged**